### PR TITLE
Optimisations - Round 3

### PR DESCRIPTION
**Pull request by lquerel**
This PR is a massive update of the underlying infrastructure used to encode/decode OTLP to/from OTLP Arrow messages.
In the previous version the `AdaptiveSchema` was used to dynamically adapt the schema for fields that can be either a dictionary or a primitive/binary type (e.g. Int32, String, Binary). 

This PR introduce a more general system called `RecordBuilderExt` (ext for extension to the standard Arrow RecordBuilder). The `RecordBuilderExt` interprets special annotations attached to the Arrow Schema of OTEL entities. There are currently 2 types of annotation supported:
* optional: used to qualify a field as optional
* dictionary: used to qualify a field as a dynamic dictionary able to detect index overflow and react accordingly either by selecting a better type for the index or by fall-backing to the value type of the dictionary.

This modification aims to improve the performance (CPU, memory allocation, compression) of small batches (<100) and to some extent improve a bit the performance of bigger batches.

Note: On the decoder side, we treat all the fields as optional as in protobuf. Logically some fields are mandatory but in fact, at the protobuf level, nothing prevents them from being optional. A default value will be assigned to any missing field (as in pdata).

A new set of unit tests have been added to track dictionary overflow for metrics, logs, and traces.
 
Updated benchmark results have been included into the README.md file.

### CODE REVIEW

File: pkg/otel/arrow_record/arrow_record.go [hunk 1]
```diff
@@ -78,13 +78,19 @@ func (rm *RecordMessage) PayloadType() PayloadType {
 func (rm *RecordMessage) ShowStats() {
 	schema := rm.record.Schema()
 	columns := rm.record.Columns()
+
+	println()
```
Comment (jmacd) [inline pkg/otel/arrow_record/arrow_record.go:82]:
```suggestion
```

File: pkg/otel/arrow_record/producer.go [hunk 2]
```diff
@@ -29,6 +29,9 @@ import (
 
 	colarspb "github.com/f5/otel-arrow-adapter/api/collector/arrow/v1"
 	acommon "github.com/f5/otel-arrow-adapter/pkg/otel/common/arrow"
+	"github.com/f5/otel-arrow-adapter/pkg/otel/common/schema"
+	"github.com/f5/otel-arrow-adapter/pkg/otel/common/schema/builder"
+	builder2 "github.com/f5/otel-arrow-adapter/pkg/otel/common/schema/config"
```
Comment (jmacd) [inline pkg/otel/arrow_record/producer.go:34]:
```suggestion
	"github.com/f5/otel-arrow-adapter/pkg/otel/common/schema/config"
```

File: pkg/otel/arrow_record/producer.go [hunk 3]
```diff
@@ -104,35 +108,31 @@ func NewProducerWithOptions(options ...Option) *Producer {
...
-		acommon.WithDictLimitIndexSize(cfg.limitIndexSize))
+		schema.WithDictInitIndexSize(cfg.initIndexSize),
+		schema.WithDictLimitIndexSize(cfg.limitIndexSize))
 
-	logsSchema := acommon.NewAdaptiveSchema(
-		cfg.pool,
-		logsarrow.Schema,
-		acommon.WithDictInitIndexSize(cfg.initIndexSize),
-		acommon.WithDictLimitIndexSize(cfg.limitIndexSize))
+	logsRecordBuilder := builder.NewRecordBuilderExt(cfg.pool, logsarrow.Schema, &builder2.DictionaryConfig{
```
Comment (jmacd) [inline pkg/otel/arrow_record/producer.go:117]: Recommend using `config.Dictionary` vs. `builder2.DictionaryConfig`. Linters will probably complain about `config.DictionaryConfig`, too.

File: pkg/otel/common/otlp2/doc.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package otlp contains common types and functions used to convert OTLP Arrow entities into their OTLP representation.
+package otlp2
```
Comment (jmacd) [inline pkg/otel/common/otlp2/doc.go:19]: I take it otlp2 and arrow2 will replace otlp and arrow? :+1:

File: pkg/otel/common/schema/builder/int.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+		return
+	}
+
+	if value != 0 {
+		// If the builder is nil, then the transform node is not optional.
+		b.transformNode.RemoveOptional()
+		b.updateRequest.Inc()
+	}
+}
```
Comment (jmacd) [inline pkg/otel/common/schema/builder/int.go:103]: I find this difficult to read, maybe a comment or explaination would help. I wonder why we're calling RemoveOptional _after_ the potential call to Append above. Don't we lose a value by skipping the first Append()?

File: pkg/otel/arrow_record/producer.go (hunk 2 above)
Comment (lquerel) [reply pkg/otel/arrow_record/producer.go:34]: Refactored the code accordingly

File: pkg/otel/arrow_record/producer.go (hunk 3 above)
Comment (lquerel) [reply pkg/otel/arrow_record/producer.go:113]: Refactored the code to use config.Dictionary.

File: pkg/otel/common/otlp2/doc.go (hunk 4 above)
Comment (lquerel) [reply pkg/otel/common/otlp2/doc.go:19]: Yes ultimately I will rename otlp2 to otlp. Same thing for arrow2.

File: pkg/otel/common/schema/builder/int.go (hunk 5 above)
Comment (lquerel) [reply pkg/otel/common/schema/builder/int.go:103]: I added a comment at the package level and on the RemoveOptional method. 

The general idea is the following. A call to `Append` data in the RecordBuilderExt must be done inside a loop in order to repeat the `Append` operation if an error of type ErrSchemaNotUpToDate occurs. This iteration will usually not exceed 1 or 2 iterations.

**Review by jmacd — APPROVED** [review]
I understand now how schema building and transformation works at the high level. Looks good. I noticed a few places where `panic` makes sense now. I filed https://github.com/f5/otel-arrow-adapter/issues/121 to make sure the collector won't die in case of panics.

File: pkg/otel/common/schema/transform_node.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+		node.Children = append(node.Children, newTransformNodeFrom(
+			path,
+			&keyField,
+			dictConfig,
+			dictTransformNodes,
+			schemaUpdateRequest,
+			events,
+		))
+
+		// ToDo remove this workaround once the arrow library supports map types with metadata (see https://github.com/apache/arrow/issues/34186).
```
[the remark this answers is not in the document]
Comment (lquerel) [reply pkg/otel/common/schema/transform_node.go:178]: I will remove this workaround in my next PR.

**Pull request merged**